[FIX] sale,website_sale: Display extra prices with taxes included

nboulif opened the pull request:
Steps to reproduce the issue:

  - Install eCommerce and Inventory module
  - Go to Settings
  - Ensure `Product Prices` is set to 'Tax included'
  - Ensure `Variant Grid Entry` is activated
  - Create a new Product as storable
  - Set `Sales Price` to $1.0
  - Set `Customer Taxes` to 10.00 %
  - Under `Variants` tab, set 'Sales Variant Selection' to 'Order Grid Entry'
  - Under `Variants` tab add an Attribute with 2 values
  - Set "Price Extra" to $2.0 to one of the variants
  - Go to the Shop and select the product

Issue:

  The extra price badge does not include the taxes
  ($2.0 instead of $2.2), however the final price does.

Cause:

  The price_extra field from ptav (used in the badge) does not include
  the taxes.
  However, when calculating the final price, we do first the sum of all
  prices (including the extra prices) and then apply the taxes.

Solution:

  In the template, for each 'variants', we call `_get_combination_info`
  to get the variant.price_extra with taxes included/excluded
  depending the `Product Prices` setting.

opw-2669871

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/80460).

nboulif commented:
Hello @tivisse :slightly_smiling_face: May I ask for a review please? Thanks in advance!

tivisse commented:
@rdeodoo Could we have your input over here ? 

At the first sight, it looks correct. But I'm wondering if the current behavior is intended or not. It seems strange that the issue is present since the 13.0 and nobody noticed it yet. 

What do you think ? 

Thanks :)

rdeodoo reviewed (approved):
This seems fine to me.

I was a bit concern about the `no_variant_attributes_price_extra` inside the `if product:` and the comment:
```python
# We need to add the price_extra for the attributes that are not
# in the variant, typically those of type no_variant, but it is
# possible that a no_variant attribute is still in a variant if
# the type of the attribute has been changed after creation.
```

But as far as I can tell, this was prevented by https://github.com/odoo/odoo/commit/054baf94a68ab1d080316469756bfba1e6a35777 so it seems ok to ignore that case.

About the commit' steps to reproduce, it seems overkill as:
- you don't need the inventory module
- you don't need to enable matrix grid
- you don't need to enable Order Grid Entry on product

So it could be reduced a bit. It is not really a problem if you keep it like that but then please swap those 2 lines:
```
  - Under `Variants` tab, set 'Sales Variant Selection' to 'Order Grid Entry'
  - Under `Variants` tab add an Attribute with 2 values
```
as you can't select Order Grid Entry without adding an attribute first, as that option is not shown if there is no attribute, misleading when you try to reproduce and don't find the option :grimacing: 
You also have to manually install Sales Management apparently to see this steps `Ensure Variant Grid Entry is activated`

> It seems strange that the issue is present since the 13.0 and nobody noticed it yet.

Well honestly it is hard to tell, because:
- There have been so many fixes about prices, including pricelist, taxes, ptav combination, price extra, discounts, currency etc
- Some of those fixes were not coming with a test, when IMHO fixes about price/numbers should always come with a test.. (and I am one to blame here too, even if that was a long time ago https://github.com/odoo/odoo/pull/22553)

Speaking about tests, could you add one?

rdeodoo commented:
> But I'm wondering if the current behavior is intended or not.

I doubt it is in this case, even if in some cases it might be, according to the comment inside the template.

> price_extra is displayed as catalog price instead of
price after pricelist because it is impossible to
compute. Indeed, the pricelist rule might depend on the
selected variant, so the price_extra will be different
depending on the selected combination. **The price of an
attribute is therefore variable and it's not very
accurate to display it**.

nboulif commented:
Hello @rdeodoo @tivisse , thanks for the review!
- Glad you told me to do the test, I noticed that I did not take into account the currency. I update also the fix for, but not sure if correct.
- I updated the existing tests since there were already many tests on `_get_combination_info`, is it enough/good?
- While I updated the tests, I noticed another issue  (here is the related test :  [test](https://github.com/odoo/odoo/blob/26b87355c42b4fd3110b6d95212b81a8d6d0a2ee/addons/website_sale/tests/test_website_sale_product_attribute_value_config.py#L94) ), but not sure related to this issue directly (or if really an issue):
  When having a `product.pricelist.item` with a `fixed` price on a product or variant, the final price does not take into account the `price_extra` but still display the price_extra on the shop product page. 
  Not sure if it is possible to fix it `easily` since must do something like [this](https://github.com/odoo/odoo/blob/14a46b64512c6b9cd8c8459aeda2fb5bf09b345e/addons/product/models/product_pricelist.py#L122) (who call [that](https://github.com/odoo/odoo/blob/14a46b64512c6b9cd8c8459aeda2fb5bf09b345e/addons/product/models/product_pricelist.py#L581)) to calculate the `price_extra`.
And If not an issue, then still must check for each variant if used in a pricelist item rule to set `price_extra=0` to avoid displaying it.

rdeodoo commented:
If I understand correctly, you are speaking about the case which is impossible to solve:
- Product: 500€
- Color White: +40€
- Color Red: +50€
- Size 128Go: +100€
- Size 256Go: +200€
- Fixed price for variant White 128Go = 600€ (instead of 640 -> 500 + 100 + 40)

Then, there is no way to show a "+ XX" on a variant attribute, as we know the variant is +100€ in total but we don't know which attribute cost what amount.
So, unless you are in "Variant list view" mode (can be enable through the customize options), it will show +100 on 128Go and +40 on White attribute while if you select those 2, the price will just be 100 as according to the variant fixed price.

That's a known limitation.

>  then still must check for each variant if used in a pricelist item rule to set price_extra=0 to avoid displaying it.

That could be an improvement, IMO for master.

nboulif commented:
@rdeodoo Thanks for the clarification :slightly_smiling_face: 
@tivisse Can you confirm that the flow and the test are correct pls? Thanks in advance

rdeodoo commented on addons/product/models/product.py:
```diff
@@ -263,9 +263,28 @@ def _set_product_lst_price(self):
+            if pricelist_id_or_name:
+                pricelist = None
+
+                # Support context pricelists specified as list, display_name or ID for compatibility
```
will this be removed in master ? (or prior?)

rdeodoo commented on addons/product/models/product.py:
```diff
@@ -263,9 +263,28 @@ def _set_product_lst_price(self):
+                pricelist = None
+
+                # Support context pricelists specified as list, display_name or ID for compatibility
+                if isinstance(pricelist_id_or_name, list):
+                    pricelist_id_or_name = pricelist_id_or_name[0]
+                if isinstance(pricelist_id_or_name, str):
+                    pricelist_name_search = self.env['product.pricelist'].name_search(pricelist_id_or_name, operator='=', limit=1)
+                    if pricelist_name_search:
+                        pricelist = self.env['product.pricelist'].browse([pricelist_name_search[0][0]])
+                elif isinstance(pricelist_id_or_name, int):
+                    pricelist = self.env['product.pricelist'].browse(pricelist_id_or_name)
```
Can't we make a private util method for this? This is exactly the same code as in `_compute_product_price()`.
Will be easier to read those methods, and easier to get rid in master (prior?).

nboulif replied:
Same here :+1:

rdeodoo commented on addons/sale/models/product_template.py:
```diff
@@ -239,8 +239,17 @@ def _get_combination_info(self, combination=False, product_id=False, add_qty=1,
             price = product.price if pricelist else list_price
             display_image = bool(product.image_1920)
             display_name = product.display_name
+            if no_variant_attributes_price_extra and pricelist:
```
Is this just for the tests to go through? Having `no_variant_attributes_price_extra` shouldn't be possible, no?
If so, definitely worth a comment to explain that IMHO.

As I mentioned in my previous message: 
> But as far as I can tell, this was prevented by 054baf9 so it seems ok to ignore that case.

nboulif replied:
It is possible to have a `no_variant_attributes_price_extra` : when you have a variant attribute with `Variants Creation Mode` set to `Never` and a `price_extra` set.

Feyensv reviewed (changes requested):
Way too big and hacky for stable IMHO.

Even more for a fix in an old version, only available if the code of `sale` is updated (risky for some existing instances, even more for old instances) or for new instances.

At least the modification of `_compute_price_extra` should be avoided.
And I fear you might be opening other complex cases.

There is a whole js code logic to update the combination info according to customer actions, and this logic isn't included in your `price_extra` changes, opening the way for even more bugfixes.

If you start considering the extra prices to be reflecting precisely the price change after selection, I fear you might have to consider the pricelist discounts (potentially depending on qty change), ...

I highly doubt we want to go that far with bugfixes in stable versions, considering the current complexity of this whole logic.

nboulif commented on addons/website_sale/tests/test_website_sale_product_attribute_value_config.py:
```diff
@@ -76,6 +79,14 @@ def test_get_combination_info_with_fpos(self):
+            'value_ids': [(6, 0, [self.ssd_256.id])],
+        })
+        computer_ssd_attribute_lines.product_template_value_ids[0].price_extra = 200
+        combinaison = computer_ssd_attribute_lines.product_template_value_ids[0]
```
My bad :sweat_smile: thanks

nboulif commented on addons/product/models/product.py:
```diff
@@ -263,9 +263,28 @@ def _set_product_lst_price(self):
             product.write({'list_price': value})
 
+    @api.depends_context('pricelist')
     def _compute_product_price_extra(self):
         for product in self:
-            product.price_extra = sum(product.product_template_attribute_value_ids.mapped('price_extra'))
+            price_extra = sum(product.product_template_attribute_value_ids.mapped('price_extra'))
+            pricelist_id_or_name = self._context.get('pricelist')
+            if pricelist_id_or_name:
+                pricelist = None
+
```
Alright, I updated the code like you suggest in MP, thanks again

nboulif commented:
Hello @Feyensv , I updated the code, do you think it's ok like that? Thanks in advance!
cc @tivisse @rdeodoo

Feyensv commented on addons/sale/models/product_template.py:
```diff
@@ -254,6 +257,10 @@ def _get_combination_info(self, combination=False, product_id=False, add_qty=1,
                 list_price, pricelist.currency_id, product_template._get_current_company(pricelist=pricelist),
                 fields.Date.today()
             )
+            price_extra = product_template.currency_id._convert(
+                price_extra, pricelist.currency_id, product_template._get_current_company(pricelist=pricelist),
+                fields.Date.today()
+            )
```
Shouldn't the currency conversion be outside of the else part to convert the product price_extra and the template price_extra ?
Also, you should probably avoid doing the currency conversion if the currencies are the same, to avoid unwanted rounding changes :thinking:

tivisse replied:
:+1:

Feyensv replied:
Forget what I said, it's already in a separate if checking the currency diff :facepalm:

tivisse commented:
@robodoo r+

rdeodoo commented:
@nboulif :point_up:
